src/useImageDimensions.ts

```diff
@@ -0,0 +1,59 @@
+import {useEffect, useState} from 'react'
+import {Image, ImageRequireSource} from 'react-native'
+
+export interface URISource {
+  uri: string
+}
+
+type ImageDimensions = {
+  width: number
+  height: number
+} | null
+type FailureCallback = (error: any) => void
+
+/**
+ * @param source local file resource.
+ * @returns Original image width and height.
+ */
+function useImageDimensions(source: ImageRequireSource): ImageDimensions
+/**
+ * @param source remote URL
+ * @param failure the function that will be called if there was an error, such as failing to retrieve the image (see https://reactnative.dev/docs/image#getsize).
+ * @returns Original image width and height.
+ */
+function useImageDimensions(
+  source: URISource,
+  failure?: FailureCallback,
+): ImageDimensions
+function useImageDimensions(
+  source: ImageRequireSource | URISource,
+  failure?: FailureCallback,
+): ImageDimensions
+function useImageDimensions(
+  source: ImageRequireSource | URISource,
+  failure?: FailureCallback,
+) {
```

Comment:
```suggestion
function useImageDimensions(
  source: ImageRequireSource | URISource,
  failure?: FailureCallback,
): ImageDimensions {
```

Is there a reason these cannot be collapsed into simply one declaration? 🤔

Reply:
IntelliSense works better with such function polymorphism
<img width="410" alt="Image 2020-03-19 at 2 53 15 AM" src="https://user-images.githubusercontent.com/44202486/77017922-d817d580-698c-11ea-8edc-7c5cc5bb10c8.png">

<img width="445" alt="Image 2020-03-19 at 2 50 32 AM" src="https://user-images.githubusercontent.com/44202486/77017887-b4548f80-698c-11ea-8174-6bb8f2bd638c.png">